File: problems/arrays-hashing/group-anagrams/solution_test.go
package group_anagrams

import (
	"reflect"
	"sort"
	"testing"
)

// sortGroups sorts groups for comparison (order of groups doesn't matter)
func sortGroups(groups [][]string) [][]string {
	// Sort strings within each group
	for i := range groups {
		sort.Strings(groups[i])
	}

	// Sort groups by their first element
	sort.Slice(groups, func(i, j int) bool {
		if len(groups[i]) == 0 || len(groups[j]) == 0 {
			return len(groups[i]) < len(groups[j])
		}
		return groups[i][0] < groups[j][0]
	})

	return groups
}

func TestGroupAnagrams(t *testing.T) {
	tests := []struct {
		name     string
		input    []string
		expected [][]string
	}{
		{
			name:  "Example 1: Basic case with multiple anagram groups",
			input: []string{"eat", "tea", "tan", "ate", "nat", "bat"},
			expected: [][]string{
				{"bat"},
				{"nat", "tan"},
				{"ate", "eat", "tea"},
			},
		},
		{
			name:     "Example 2: Single empty string",
			input:    []string{""},
			expected: [][]string{{""}},
		},
		{
			name:     "Example 3: Single character string",
			input:    []string{"a"},
			expected: [][]string{{"a"}},
		},
		{
			name:     "Edge case: empty array",
			input:    []string{},
			expected: [][]string{},
		},
		{
			name:  "Edge case: no anagrams (all unique words)",
			input: []string{"abc", "def", "ghi"},
			expected: [][]string{
				{"abc"},
				{"def"},
				{"ghi"},
			},
		},
		{
			name:  "Edge case: all strings are anagrams",
			input: []string{"abc", "bca", "cab", "acb"},
			expected: [][]string{
				{"abc", "bca", "cab", "acb"},
			},
		},
		{
			name:  "Complex case: multiple groups with different lengths",
			input: []string{"listen", "silent", "hello", "world", "enlist"},
			expected: [][]string{
				{"hello"},
				{"listen", "silent", "enlist"},
				{"world"},
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := GroupAnagrams(tt.input)

			// Sort both result and expected for comparison
			sortedResult := sortGroups(result)
			sortedExpected := sortGroups(tt.expected)

			if !reflect.DeepEqual(sortedResult, sortedExpected) {
				t.Errorf("GroupAnagrams() = %v, want %v", sortedResult, sortedExpected)
			}
		})
	}
}

File: problems/arrays-hashing/group-anagrams/solution.test.ts
import { describe, test, expect } from 'vitest';
import { groupAnagrams } from './solution';

// Helper function to sort arrays for comparison (order of groups doesn't matter)
function sortGroups(groups: string[][]): string[][] {
  return groups
    .map(group => group.sort())
    .sort((a, b) => a[0].localeCompare(b[0]));
}

describe('Group Anagrams', () => {
  test('Example 1: Basic case with multiple anagram groups', () => {
    const input = ["eat", "tea", "tan", "ate", "nat", "bat"];
    const result = groupAnagrams(input);
    const expected = [["bat"], ["nat", "tan"], ["ate", "eat", "tea"]];

    expect(sortGroups(result)).toEqual(sortGroups(expected));
  });

  test('Example 2: Single empty string', () => {
    const input = [""];
    const result = groupAnagrams(input);
    const expected = [[""]];

    expect(result).toEqual(expected);
  });

  test('Example 3: Single character string', () => {
    const input = ["a"];
    const result = groupAnagrams(input);
    const expected = [["a"]];

    expect(result).toEqual(expected);
  });

  test('Edge case: empty array', () => {
    const input: string[] = [];
    const result = groupAnagrams(input);
    const expected: string[][] = [];

    expect(result).toEqual(expected);
  });

  test('Edge case: no anagrams (all unique words)', () => {
    const input = ["abc", "def", "ghi"];
    const result = groupAnagrams(input);
    const expected = [["abc"], ["def"], ["ghi"]];

    expect(sortGroups(result)).toEqual(sortGroups(expected));
  });

  test('Edge case: all strings are anagrams', () => {
    const input = ["abc", "bca", "cab", "acb"];
    const result = groupAnagrams(input);

    expect(result.length).toBe(1);
    expect(result[0].sort()).toEqual(["abc", "acb", "bca", "cab"]);
  });

  test('Complex case: multiple groups with different lengths', () => {
    const input = ["listen", "silent", "hello", "world", "enlist"];
    const result = groupAnagrams(input);
    const expected = [["hello"], ["listen", "silent", "enlist"], ["world"]];

    expect(sortGroups(result)).toEqual(sortGroups(expected));
  });
});
